daemon/api_model_test.go

```diff
@@ -499,3 +504,137 @@ func (s *userSuite) TestPostSerialForgetError(c *check.C) {
+	s.expectRootAccess()
+
+	oldModel := s.Brands.Model("my-brand", "my-old-model", modelDefaults)
+	newModel := s.Brands.Model("my-brand", "my-old-model", modelDefaults, map[string]interface{}{
+		"revision": "2",
+	})
+
+	d := s.daemonWithOverlordMockAndStore()
+	hookMgr, err := hookstate.Manager(d.Overlord().State(), d.Overlord().TaskRunner())
+	c.Assert(err, check.IsNil)
+	deviceMgr, err := devicestate.Manager(d.Overlord().State(), hookMgr, d.Overlord().TaskRunner(), nil)
+	c.Assert(err, check.IsNil)
+	d.Overlord().AddManager(deviceMgr)
+	st := d.Overlord().State()
+	st.Lock()
+	assertstatetest.AddMany(st, s.StoreSigning.StoreAccountKey(""))
+	assertstatetest.AddMany(st, s.Brands.AccountsAndKeys("my-brand")...)
+	s.mockModel(st, oldModel)
+	st.Unlock()
+
+	soon := 0
+	var origEnsureStateSoon func(*state.State)
+	origEnsureStateSoon, restore := daemon.MockEnsureStateSoon(func(st *state.State) {
+		soon++
+		origEnsureStateSoon(st)
+	})
+	defer restore()
+
+	snapName := "snap1"
+	snapRev := 1001
+	var devicestateRemodelGotModel *asserts.Model
+	defer daemon.MockDevicestateRemodel(func(st *state.State, nm *asserts.Model,
+		sis []*snap.SideInfo, paths []string) (*state.Change, error) {
+		c.Check(len(sis), check.Equals, 1)
+		c.Check(sis[0].RealName, check.Equals, snapName)
+		c.Check(sis[0].Revision, check.Equals, snap.Revision{N: snapRev})
+		c.Check(strings.HasSuffix(paths[0],
+			"/var/lib/snapd/snaps/"+snapName+"_"+strconv.Itoa(snapRev)+".snap"),
+			check.Equals, true)
+
+		devicestateRemodelGotModel = nm
+		chg := st.NewChange("remodel", "...")
+		return chg, nil
+	})()
+
+	sis := []*snap.SideInfo{{RealName: snapName, Revision: snap.Revision{N: snapRev}}}
+	defer daemon.MockSideloadSnapsInfo(sis)()
+
+	// create a valid model assertion
+	c.Assert(err, check.IsNil)
+	var modelEncoded string
+	if params.badModel {
+		modelEncoded = "garbage"
+	} else {
+		modelEncoded = string(asserts.Encode(newModel))
+	}
+
+	// valid revision assertion to make it part of the arguments
+	revAssert := assertstest.FakeAssertion(map[string]interface{}{
+		"type":          "snap-revision",
+		"authority-id":  "can0nical",
+		"snap-id":       "snap-id-1",
+		"snap-sha3-384": "1111111111111111111111111111111111111111111111111111111111111111",
+		"snap-size":     fmt.Sprintf("%d", 100),
+		"snap-revision": strconv.Itoa(snapRev),
+		"developer-id":  "mydev",
+		"timestamp":     time.Now().Format(time.RFC3339),
+	}).(*asserts.SnapRevision)
+
+	// create multipart data
+	body, boundary := multipartBody(c, modelEncoded, "snap_data", string(revAssert.Body()))
+
+	// set it and validate that this is what we passed to devicestateRemodel
+	req, err := http.NewRequest("POST", "/v2/model", &body)
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/form-data; boundary="+boundary)
+	req.Header.Set("Content-Length", strconv.Itoa(body.Len()))
+
+	if params.badModel {
```

Comment:
nitpick: i'm not a total fan of doing unit tests this way, I believe this function should instead return whatever the unit test needs, and then the individual unit tests perform their validation of the result. My personal preference would be to move whatever is in this if/else into their respective tests, and then keep the common code for the unit test as clean as possible to also allow for easier reuse in the future.